applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.cpp

```diff
@@ -225,6 +225,39 @@ double OptimizationUtilities::ComputeCorrectionFactor(ModelPart& rModelPart, con
     return CorrectionScaling * norm_search_direction / norm_correction_term;
 }
 
+void OptimizationUtilities::AssembleScalar( ModelPart& rModelPart,
+    Vector& rScalar,
+    const Variable<double> &rVariable)
+{
+    if (rScalar.size() != rModelPart.NumberOfNodes()){
+        rScalar.resize(rModelPart.NumberOfNodes());
+    }
+
+    int i=0;
+    for (auto & node_i : rModelPart.Nodes())
+    {
+        rScalar[i] = node_i.FastGetSolutionStepValue(rVariable);
+        ++i;
+    }
+}
+
+void OptimizationUtilities::AssignScalarToVariable(ModelPart& rModelPart,
+    const Vector& rScalar,
+    const Variable<double> &rVariable)
+{
+    KRATOS_ERROR_IF(rScalar.size() != rModelPart.NumberOfNodes())
+        << "AssignScalarToVariable: Scalar size does not mach number of Nodes!" << std::endl;
+
+    int i=0;
+    for (auto & node_i : rModelPart.Nodes())
+    {
+        double& variable_vector = node_i.FastGetSolutionStepValue(rVariable);
+        variable_vector= rScalar[i];
+        ++i;
+    }
+}
+
+
```

Comment:
those function exist (more efficient even) in the AuxiliarModelPartUtilities

Reply:
thanks a lot for all the suggestions.

I am struggling with directly using the methods from AuxiliarModelPartUtilities because they expect `std::vector`, but I am using `KratosMultiphysics.Vector()` coming from the python side. 

So I rewrote the utility function to:
```
void OptimizationUtilities::AssembleVector( ModelPart& rModelPart,
    Vector& rVector,
    const Variable<double> &rVariable)
{

    std::vector<double> data;
    AuxiliarModelPartUtilities(rModelPart).GetScalarData<double>(rVariable, DataLocation::NodeHistorical, data);
    if (rVector.size() != rModelPart.NumberOfNodes()){
        rVector.resize(rModelPart.NumberOfNodes());
    }
    for (int i = 0; i < data.size(); ++i)
    {
        rVector[i] = data[i];
    }
}
```

Which is far from optimal with respect do the data handling. How could that be improved?

Reply:
Hey,

I think I can improve the fct to also work with Vector.
I should have time for it at some point next week if you can wait that long :)

Reply:
That would be great :) If it is early next week, should be okay. 

But I guess that this is not a blocker, right? So I could change those functions also in a follow-up pr.

Reply:
sure, no need to wait

with #10548 it should work, feel free to try :)

Reply:
I've changed the code to fully utilize Kratos.Vector! All variables are accessed now using Vector and the respective heatmaps are also computed by Vector